### DISCUSSION TITLE
Gulp-typescript out of memory error

### DISCUSSION THREAD

**Discussion by skhilliard** [actions/virtual-environments#4351] [Image configuration]
I recently tried to update an package for my Node.js project (mongoose 5.3.12), and it seems to have pushed me over the edge for memory during the typescript compilation phase (using gulp-typescript 5.0.1).  Here is the error I see...

[18:24:13] Starting 'node_ts'...

<--- Last few GCs --->

[480:0x58548e0]   225969 ms: Mark-sweep (reduce) 2040.9 (2051.1) -> 2039.9 (2051.8) MB, 2662.0 / 0.0 ms  (average mu = 0.095, current mu = 0.003) allocation failure scavenge might not succeed
[480:0x58548e0]   225977 ms: Scavenge (reduce) 2040.9 (2050.8) -> 2040.0 (2051.8) MB, 2.9 / 0.0 ms  (average mu = 0.095, current mu = 0.003) allocation failure 

<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0xa389b0 node::Abort() [gulp default --semVer=21.11.0.72]
 2: 0x96e0af node::FatalError(char const*, char const*) [gulp default --semVer=21.11.0.72]
 3: 0xbb7a4e v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [gulp default --semVer=21.11.0.72]
 4: 0xbb7dc7 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [gulp default --semVer=21.11.0.72]
 5: 0xd73fd5  [gulp default --semVer=21.11.0.72]
 6: 0xd74b5f  [gulp default --semVer=21.11.0.72]
 7: 0xd8299b v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [gulp default --semVer=21.11.0.72]
 8: 0xd8655c v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [gulp default --semVer=21.11.0.72]
 9: 0xd54c3b v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationType, v8::internal::AllocationOrigin) [gulp default --semVer=21.11.0.72]
10: 0x109d21f v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [gulp default --semVer=21.11.0.72]
11: 0x1446379  [gulp default --semVer=21.11.0.72]
Aborted (core dumped)
npm ERR! code ELIFECYCLE
npm ERR! errno 134
npm ERR! evolved@1.0.0 build-prod: `gulp default "--semVer=21.11.0.72"`

I am sure this is newbie question, but do I have options to increase available memory?  From what I understand this docker image has 6 GB of physical memory and 13 GB of total memory. 

Any suggestions or advice would be greatly appreciated.

Thanks,
Kelly

**Comment by miketimofeev** [comment]
Hey @skhilliard!
What docker image do you mean? Our hosted VMs has 7Gb of total ram.
I think you can try to increase the max heap, something like this https://stackoverflow.com/questions/38558989/node-js-heap-out-of-memory

**Reply by skhilliard** [reply]
We are using node:14-slim as our base image...if I understand correctly the "build" is happening in that image (which is hosted in your hosted VM).

We do issue a command like what is referenced above as part of the build....not sure if it applies for every command after or not.  Here is some more of the output that shows that...do we need to apply that differently somehow for it to apply to the Typescript compilation itself?  The node_ts step is executed as a task in Gulpfile.js using the gulp-typescript plugin.

[18:22:01] Using gulpfile /server/Gulpfile.js
[18:22:01] Starting 'default'...
[18:22:01] Starting 'clean'...
[18:22:01] Finished 'clean' after 73 ms
[18:22:01] Starting 'version'...
[18:22:04] Finished 'version' after 2.62 s
[18:22:04] Starting 'eslint'...

> evolved@1.0.0 lint /server
> **node --max_old_space_size=4096 ./node_modules/eslint/bin/eslint.js . --ext .ts**

[18:24:13] Finished 'eslint' after 2.15 min
**[18:24:13] Starting 'node_ts'...**

<--- Last few GCs --->

[480:0x58548e0]   225969 ms: Mark-sweep (reduce) 2040.9 (2051.1) -> 2039.9 (2051.8) MB, 2662.0 / 0.0 ms  (average mu = 0.095, current mu = 0.003) allocation failure scavenge might not succeed
[480:0x58548e0]   225977 ms: Scavenge (reduce) 2040.9 (2050.8) -> 2040.0 (2051.8) MB, 2.9 / 0.0 ms  (average mu = 0.095, current mu = 0.003) allocation failure 


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0xa389b0 node::Abort() [gulp default --semVer=21.11.0.72]
 2: 0x96e0af node::FatalError(char const*, char const*) [gulp default --semVer=21.11.0.72]
 3: 0xbb7a4e v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [gulp default --semVer=21.11.0.72]
 4: 0xbb7dc7 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [gulp default --semVer=21.11.0.72]
 5: 0xd73fd5  [gulp default --semVer=21.11.0.72]
 6: 0xd74b5f  [gulp default --semVer=21.11.0.72]
 7: 0xd8299b v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [gulp default --semVer=21.11.0.72]
 8: 0xd8655c v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [gulp default --semVer=21.11.0.72]
 9: 0xd54c3b v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationType, v8::internal::AllocationOrigin) [gulp default --semVer=21.11.0.72]
10: 0x109d21f v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [gulp default --semVer=21.11.0.72]
11: 0x1446379  [gulp default --semVer=21.11.0.72]
Aborted (core dumped)
npm ERR! code ELIFECYCLE
npm ERR! errno 134
npm ERR! evolved@1.0.0 build-prod: `gulp default "--semVer=21.11.0.72"`
npm ERR! Exit status 134
npm ERR! 
npm ERR! Failed at the evolved@1.0.0 build-prod script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

**Reply by skhilliard** [reply]
FYI....I've tried setting NODE_OPTIONS thusly at the beginning of the build to get around this...

NODE_OPTIONS=--max-old-space-size=6144

...but it doesn't seem to have helped.

**Reply by miketimofeev** [reply]
Does this work for you locally, outside of GH actions?

**Reply by miketimofeev** [reply]
You can also play around with the options for the container https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#jobsjob_idcontaineroptions, there is a memory option at least

**Reply by skhilliard** [reply]
Yes....I can build locally (on my Windows box) with Visual Studio Code with no issues at all.  The build is failing in the Azure Pipeline only.

**Reply by skhilliard** [reply]
Just FYI....I added  this command to the build scripts to validate how much memory is available...

grep MemTotal /proc/meminfo

...and it did report 7 GB was there, which is consistent with your comment, @miketimofeev .

I am not sure why setting the environment variable...

NODE_OPTIONS=--max-old-space-size=6144

...didn't work.  Are there other things that have to be done to have memory available to tsc?

**Reply by dmitry-shibanov** [reply]
Hello @skhilliard. Could you please create an [issue](https://github.com/actions/virtual-environments/issues/new/choose) and provide public repository to investigate dipper or simple repro steps to reproduce the issue ?

**Reply by skhilliard** [reply]
@dmitry-shibanov Sorry for the late reply, but have been busy with moving and then the holidays hit.  I did as you suggested and created an issue:  https://github.com/actions/virtual-environments/issues/4649

Let me know what other information I can provide to help.

Thanks,
Kelly

**Comment by samus-aran** [comment]
We're locking this discussion because it has not had recent activity and/or other members have asked for more information to assist you but received no response. Thank you for helping us maintain a productive and tidy community for all our members.
